src/powerquery-formatter/serializeV2.ts

```diff
@@ -0,0 +1,566 @@
+            const maybeChildren: ReadonlyArray<Ast.TNode> | undefined =
+                PQP.Parser.NodeIdMapIterator.maybeIterChildrenAst(state.nodeIdMapCollection, node.id);
+
+            if (maybeChildren === undefined) {
+                break;
+            }
+
+            for (const child of maybeChildren) {
+                // eslint-disable-next-line no-await-in-loop
+                await serializeNode(state, child, { isParentInline: currentlySupportInlineBlock });
+            }
+        }
+    }
+
+    if (parameter.blockOpener === "R" || parameter.blockCloser === "R") {
+        if (parameter.blockOpener) {
+            // eslint-disable-next-line require-atomic-updates
+            state.lastTokenType = "Opener";
+
+            if (currentlySupportInlineBlock) {
+                state.blockStatusArray.push("InlineBlock");
+            } else {
+                state.blockStatusArray.push("Block");
+                appendToFormatted(state, state.newlineLiteral);
+                state.indentationLevel += 1;
+            }
+        } else if (parameter.blockCloser) {
+            const curBlockStatus: BlockStatus | undefined = state.blockStatusArray.pop();
+
+            if (curBlockStatus === "Block") {
+                state.indentationLevel -= 1;
+                appendToFormatted(state, state.newlineLiteral);
+            }
+
+            state.lastTokenType = "Closer";
+        }
+    } else if (parameter.contentDivider === "R" && state.lastTokenType !== "Divider") {
+        // eslint-disable-next-line require-atomic-updates
+        state.lastTokenType = "Divider";
+        const curBlockStatus: BlockStatus | undefined = currentBlockStatus(state);
+
+        if (curBlockStatus === "Block") {
+            appendToFormatted(state, state.newlineLiteral);
+        }
+    }
+
+    // restore indent level for a container node
+    if (isContainer) {
+        // false alarm: there should be no race problem since we serialize linearly,
+        // we did not promise all at its caller
+        // eslint-disable-next-line require-atomic-updates
+        state.indentationLevel = currentIndentLevel;
+        state.blockStatusArray = state.blockStatusArray.slice(0, currentBlockStatusArrLen);
+
+        if (!currentlySupportInlineBlock && state.currentLine !== "" && parameter.skipPostContainerNewLine !== true) {
+            appendToFormatted(state, state.newlineLiteral);
+        }
+    }
+}
+
+function getSectionMemberLiteralTailPair(sectionLiteral: string): [string, string] {
+    const dotIndex: number = sectionLiteral.indexOf(".");
+    let head: string;
+    let tail: string;
+
+    if (dotIndex === -1) {
+        head = sectionLiteral;
+        tail = "";
+    } else {
+        head = sectionLiteral.substring(0, dotIndex);
+        tail = sectionLiteral.substring(dotIndex + 1);
+    }
+
+    return [head, tail];
+}
+
+function shouldBreakAwayFromLastSectionMember(lastSectionLiteral: string, currentSectionLiteral: string): boolean {
```

Comment:
I think this never is true unless both identifiers are exactly equal, which in itself is an error as section members cannot share a name.

I think what you want is to grab the heads of both and compare them while ignoring the tails. Currently it's a tuple comparison.

Reply:
No, no, it works; it won't put a new line between the equality expressions when their shared the same first namespace name; and i just did one more change: i would regard the plain l-identifier w/o punctuation belonging to the empty string"" root namespace which would also avoiding creating unneeded new lines b/w them

Reply:
Gotcha, sounds good 👍